test/sqllogictest/transform/predicate_pushdown.slt

```diff
@@ -327,12 +327,11 @@ Explained Query:
       project=(#0, #1, #0, #2)
     linear_stage[0]
       closure
-        project=(#0, #2, #1)
         filter=((#0) IS NOT NULL AND (#3 OR #4) AND (#5 OR #6) AND ((#3 AND #5) OR (#4 AND #6)))
-        map=((#2 = 3), (#2 = 5), (#1 = 4), (#1 = 6))
-      lookup={ relation=0, key=[#0] }
+        map=((#1 = 3), (#1 = 5), (#2 = 4), (#2 = 6))
+      lookup={ relation=1, key=[#0] }
       stream={ key=[#0], thinning=(#1) }
-    source={ relation=1, key=[#0] }
+    source={ relation=0, key=[#0] }
```

Comment:
I'm not really sure what happened here, but did want to 1. discuss, but also 2. flag that the "predicate implications" logic seems to have done something suboptimal here in that 
```
        filter=((#0) IS NOT NULL AND (#3 OR #4) AND (#5 OR #6) AND ((#3 AND #5) OR (#4 AND #6)))
```
should be
```
        filter=((#0) IS NOT NULL AND ((#3 AND #5) OR (#4 AND #6)))
```

Reply:
Yes, unfortunately the "predicate implications" logic can sometimes make these redundant predicates appear in the last stage of a physical join plan. This happens as follows:
- The "predicate implications" logic looks at a predicate P, and creates an implication I, which it adds at a join input.
- `JoinImplementation` lifts away I, because it wants to use an arrangement. Now we have `P AND I` in the `Filter` that is on top of the join, making the MIR plan look weird.
- Lowering pushes down I into the closure of the stage where it was lifted from by `JoinImplementation` (to be executed _after_ executing the join that is in that stage).
  - If this is not the last stage, then the MIR plan's weirdness is not present in the physical plan: I will be executed before some other join stage, so having I makes sense.
  - If this happens to be the last stage, then this is where `P` also ends up, so in this case even the physical plan has the weirdness `P AND I`: we would be better off with `I` not existing at all.

We could fix this by having a new scalar expression optimization pass that would check each argument of an AND whether it is an implication of all the other arguments, and drop it in that case. I opened an issue for this: https://github.com/MaterializeInc/materialize/issues/16517

(Btw. we have another stale comment here, which I'm tracking in https://github.com/MaterializeInc/materialize/issues/16478#issuecomment-1339846336)